src/main/java/com/faforever/api/map/MapService.java

```diff
@@ -0,0 +1,348 @@
+    map.getVersions().add(version);
+    version.setMap(map);
+
+    progressData.setMapEntity(map);
+    progressData.setMapVersionEntity(version);
+
+    version.setFilename(progressData.getFinalZipName());
+    progressData.setFinalZipFile(Paths.get(
+        this.fafApiProperties.getMap().getFinalDirectory(),
+        progressData.getFinalZipName()));
+
+    if (Files.exists(progressData.getFinalZipFile())) {
+      throw new ApiException(new Error(ErrorCode.MAP_NAME_CONFLICT, progressData.getFinalZipName()));
+    }
+
+    // save entity to db to trigger validation
+    mapRepository.save(map);
+  }
+
+  @SneakyThrows
+  private void renameFolderNameAndAdaptLuaFiles(MapUploadData progressData) {
+    progressData.setNewMapFolder(Paths.get(progressData.getBaseDir().toString(), progressData.getNewFolderName()));
+    Files.move(progressData.getOriginalMapFolder(), progressData.getNewMapFolder());
+    updateLuaFiles(progressData);
+  }
+
+  @SneakyThrows
+  private void updateLuaFiles(MapUploadData mapData) {
+    String oldNameFolder = "/maps/" + mapData.getUploadFolderName();
+    String newNameFolder = "/maps/" + mapData.getNewFolderName();
+    try (Stream<Path> mapFileStream = Files.list(mapData.getNewMapFolder())) {
+      mapFileStream
+          .filter(path -> path.toString().toLowerCase().endsWith(".lua"))
+          .forEach(path -> noCatch(() -> {
+            List<String> collect = Files.readAllLines(path, MAP_CHARSET).stream()
+                .map(line -> line.replaceAll("(?i)" + oldNameFolder, newNameFolder))
+                .collect(Collectors.toList());
+            Files.write(path, collect, MAP_CHARSET);
+          }));
+    }
+  }
+
+
+  private void generatePreview(MapUploadData mapData) throws IOException {
+    String previewFilename = mapData.getNewFolderName() + ".png";
+    generateImage(Paths.get(
+        fafApiProperties.getMap().getMapPreviewPathSmall(), previewFilename),
+        mapData.getNewMapFolder(),
+        fafApiProperties.getMap().getPreviewSizeSmall());
+
+    generateImage(Paths.get(
+        fafApiProperties.getMap().getMapPreviewPathLarge(), previewFilename),
+        mapData.getNewMapFolder(),
+        fafApiProperties.getMap().getPreviewSizeLarge());
+  }
+
+  @SneakyThrows
+  private void zipMapData(MapUploadData progressData) {
+    cleanupBaseDir(progressData);
+    try (ZipOutputStream zipOutputStream = new ZipOutputStream(new BufferedOutputStream(
+        Files.newOutputStream(progressData.getFinalZipFile())))) {
+      Zipper.contentOf(progressData.getBaseDir()).to(zipOutputStream).zip();
+    }
+  }
+
+  private void cleanupBaseDir(MapUploadData progressData) throws IOException {
+    Files.delete(progressData.getUploadedFile());
+    try (Stream<Path> stream = Files.list(progressData.getBaseDir())) {
+      if (stream.count() != 1) {
+        throw new IllegalStateException("Folder containing unknown data: " + progressData.getBaseDir());
```

Comment:
"contains". Again, this is a user error and therefore isn't an invalid "state" but should be an API exception.